app/lib/service/openid/jwt.dart

```diff
@@ -123,7 +119,189 @@ Map<String, dynamic> _decodePart(String part) {
+  }
+  throw FormatException('Unexpected value for `$key`: `$value`.');
+}
+
+String? _parseAsStringOrNull(Map<String, dynamic> map, String key) {
+  final value = map[key];
+  if (value == null) {
+    return null;
+  }
+  if (value is String) {
+    return value;
+  }
+  throw FormatException('Unexpected value for `$key`: `$value`.');
+}
+
+String _parseAsString(Map<String, dynamic> map, String key) {
+  final value = _parseAsStringOrNull(map, key);
+  if (value == null) {
+    throw FormatException('Missing value for `$key`.');
+  } else {
+    return value;
+  }
+}
+
+/// The parsed JWT header.
+class JwtHeader extends UnmodifiableMapView<String, dynamic> {
+  /// algorithm
+  final String? alg;
+
+  /// type
+  final String? typ;
+
+  /// key identifier
+  final String? kid;
+
+  JwtHeader._(super.map)
+      : alg = _parseAsStringOrNull(map, 'alg'),
+        typ = _parseAsStringOrNull(map, 'typ'),
+        kid = _parseAsStringOrNull(map, 'kid');
+
+  factory JwtHeader(Map<String, dynamic> values) {
+    try {
+      return JwtHeader._(values);
+    } on FormatException {
+      rethrow;
+    } catch (e, st) {
+      _logger.warning('Unexpected JWT parser exception.', e, st);
+      throw FormatException('Unexpected value in JWT header.');
+    }
+  }
+}
+
+/// The parsed JWT payload.
+class JwtPayload extends UnmodifiableMapView<String, dynamic> {
+  /// timestamp when token was issued
+  final DateTime? iat;
+
+  /// timestamp before which the token is not valid
+  final DateTime? nbf;
+
+  /// timestamp at which the token expires
+  final DateTime? exp;
+
+  /// The "iss" (issuer) claim identifies the principal that issued the JWT.
+  final String? iss;
+
+  JwtPayload._(super.map)
+      : iat = _parseIntAsSecondsOrNull(map, 'iat'),
+        nbf = _parseIntAsSecondsOrNull(map, 'nbf'),
+        exp = _parseIntAsSecondsOrNull(map, 'exp'),
+        iss = _parseAsStringOrNull(map, 'iss');
+
+  factory JwtPayload(Map<String, dynamic> map) {
+    try {
+      return JwtPayload._(map);
+    } on FormatException {
+      rethrow;
+    } catch (e, st) {
+      _logger.warning('Unexpected JWT parser exception.', e, st);
+      throw FormatException('Unexpected value in JWT payload.');
+    }
+  }
+
+  /// Verifies the timestamps with to the provided [now]:
+  /// - [iat] <= [now]
+  /// - [nbf] <= [now]
+  /// - [now] <= [exp]
+  ///
+  /// Returns `false` if the current timestamp is outside of the allowed range.
+  /// If the timestamp is missing, we treat it as if it would allow the current
+  /// time.
+  bool verifyTimestamps([DateTime? now]) {
+    now ??= clock.now();
+    if (iat != null && iat!.isAfter(now)) {
+      return false;
+    }
+    if (nbf != null && nbf!.isAfter(now)) {
+      return false;
+    }
+    if (exp != null && exp!.isBefore(now)) {
+      return false;
+    }
+    return true;
+  }
+}
+
+/// Parsed payload with the payload values GitHub sends with the token.
+class GitHubJwtPayload {
```

Comment:
Let's put this in a separate file... maybe `service/openid/github.dart`

Reply:
Added a TODO for it, I'll do it in a subsequent change.